gradle-plugin/src/main/java/com/squareup/hephaestus/plugin/HephaestusPlugin.kt

```diff
@@ -117,48 +118,62 @@ open class HephaestusPlugin : Plugin<Project> {
-      // incremental builds are broken. See https://youtrack.jetbrains.com/issue/KT-38570
-      val disableIncrementalCompilationTaskProvider = project.tasks.register(
-          compileTaskName + "CheckIncrementalCompilationHephaestus",
-          DisableIncrementalCompilationTask::class.java
-      ) { task ->
-        task.pluginClasspath.from(
-            project.configurations.getByName(PLUGIN_CLASSPATH_CONFIGURATION_NAME)
-        )
-        task.incrementalSignal.set(incrementalSignal)
+    if (isAndroidProject) {
+      project.androidVariantsConfigure { variant ->
+        val compileTaskName = "compile${variant.name.capitalize(US)}Kotlin"
+        disableIncrementalCompilationAction(project, incrementalSignal, compileTaskName)
       }
+    } else {
+      // The Java plugin has two Kotlin tasks we care about: compileKotlin and compileTestKotlin.
+      disableIncrementalCompilationAction(project, incrementalSignal, "compileKotlin")
+      disableIncrementalCompilationAction(project, incrementalSignal, "compileTestKotlin")
+    }
+  }
 
-      project.tasks.named(compileTaskName, KotlinCompile::class.java) { compileTask ->
-        compileTask.dependsOn(disableIncrementalCompilationTaskProvider)
+  fun disableIncrementalCompilationAction(
+    project: Project,
+    incrementalSignal: Provider<IncrementalSignal>,
+    compileTaskName: String
+  ) {
+    // Disable incremental compilation, if the compiler plugin dependency isn't up-to-date.
+    // This will trigger a full compilation of a module using Hephaestus even though its
+    // source files might not have changed. This workaround is necessary, otherwise
+    // incremental builds are broken. See https://youtrack.jetbrains.com/issue/KT-38570
+    val disableIncrementalCompilationTaskProvider = project.tasks.register(
+        compileTaskName + "CheckIncrementalCompilationHephaestus",
+        DisableIncrementalCompilationTask::class.java
+    ) { task ->
+      task.pluginClasspath.from(
+          project.configurations.getByName(PLUGIN_CLASSPATH_CONFIGURATION_NAME)
+      )
+      task.incrementalSignal.set(incrementalSignal)
+    }
 
-        compileTask.doFirst {
-          // If the signal is set, then the plugin classpath changed. Apply the setting that
-          // DisableIncrementalCompilationTask requested.
-          val incremental = incrementalSignal.get().incremental[project.path]
-          if (incremental != null) {
-            compileTask.incremental = incremental
-          }
+    project.tasks.named(compileTaskName, KotlinCompile::class.java) { compileTask ->
+      compileTask.dependsOn(disableIncrementalCompilationTaskProvider)
+    }
 
-          compileTask.logger.info(
-              "Hephaestus: Incremental compilation enabled: ${compileTask.incremental} (compile)"
-          )
+    // We avoid a reference to the project in the doFirst.
+    val projectPath = project.path
+    val incremental = incrementalSignal.get().incremental[projectPath]
```

Comment:
You call this too early. It has to happen in the doFirst, because otherwise the task that toggles this flag didn't run, yet.